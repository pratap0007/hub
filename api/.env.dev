ENVIRONMENT="development"

POSTGRES_HOST="localhost"
POSTGRES_PORT="5432"
POSTGRES_DB="hub"
POSTGRES_USER="postgres"
POSTGRES_PASSWORD="postgres"

GH_CLIENT_ID=""
GH_CLIENT_SECRET=""
GHE_URL=""

GL_CLIENT_ID=""
GL_CLIENT_SECRET=""
GLE_URL=""

BB_CLIENT_ID=""
BB_CLIENT_SECRET=""

JWT_SIGNING_KEY=""
ACCESS_JWT_EXPIRES_IN=""
REFRESH_JWT_EXPIRES_IN=""

AUTH_BASE_URL=""

CONFIG_FILE_URL="file://../config.yaml"
