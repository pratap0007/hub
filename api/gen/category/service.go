// Code generated by goa v3.1.3, DO NOT EDIT.
//
// category service
//
// Command:
// $ goa gen github.com/tektoncd/hub/api/design

package category

import (
	"context"

	goa "goa.design/goa/v3/pkg"
)

// The category service gives category details
type Service interface {
	// Get all Categories with their tags sorted by name
	All(context.Context) (res []*Category, err error)
}

// ServiceName is the name of the service as defined in the design. This is the
// same value that is set in the endpoint request contexts under the ServiceKey
// key.
const ServiceName = "category"

// MethodNames lists the service method names as defined in the design. These
// are the same values that are set in the endpoint request contexts under the
// MethodKey key.
var MethodNames = [1]string{"All"}

type Category struct {
	// unique id of category
	ID uint
	// name of category
	Name string
	// list of tag associated with category
	Tags []*Tag
}

type Tag struct {
	// Id is the unique id of tags
	ID uint
	// name of tag
	Name string
}

// MakeInternalError builds a goa.ServiceError from an error.
func MakeInternalError(err error) *goa.ServiceError {
	return &goa.ServiceError{
		Name:    "internal-error",
		ID:      goa.NewErrorID(),
		Message: err.Error(),
	}
}
