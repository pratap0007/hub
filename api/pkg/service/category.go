package hub

import (
	"context"
	"log"

	category "github.com/tektoncd/hub/api/gen/category"
)

// category service example implementation.
// The example methods log the requests and return zero values.
type categorysrvc struct {
	logger *log.Logger
}

// NewCategory returns the category service implementation.
func NewCategory(logger *log.Logger) category.Service {
	return &categorysrvc{logger}
}

// Get all Categories with their tags sorted by name
func (s *categorysrvc) All(ctx context.Context) (res []*category.Category, err error) {
	s.logger.Print("category.All")
	return
}
