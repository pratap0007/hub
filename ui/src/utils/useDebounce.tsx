import { useEffect, useState } from 'react';
import { useMst } from '../store/root';

export const useDebounce = (value: string, delay: number) => {
  const [debouncedValue, setDebouncedValue] = useState(value);
  const { resources } = useMst();

  useEffect(() => {
    const handler = setTimeout(() => {
      setDebouncedValue(value);
      resources.setSearch(value);
    }, delay);

    return () => {
      clearTimeout(handler);
    };
  }, [value, delay, resources]);
  return debouncedValue;
};
