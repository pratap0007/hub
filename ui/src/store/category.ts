import { AxiosError } from 'axios';
import { types, getEnv, flow, Instance } from 'mobx-state-tree';
import { Api } from '../api';

export const Category = types
  .model('Category', {
    id: types.identifierNumber,
    name: types.string,
    selected: false
  })

  .actions((self) => ({
    toggle() {
      self.selected = !self.selected;
    }
  }));

export type ICategory = Instance<typeof Category>;
export type ICategoryStore = Instance<typeof CategoryStore>;

export const CategoryStore = types
  .model('CategoryStore', {
    items: types.map(Category),
    isLoading: true,
    err: ''
  })

  .views((self) => ({
    get api(): Api {
      return getEnv(self).api;
    },

    get count() {
      return self.items.size;
    },

    get values() {
      return Array.from(self.items.values());
    },

    // This returns list of selected category's name
    get selectedByName() {
      return Array.from(self.items.values())
        .filter((c: ICategory) => c.selected)
        .reduce((acc: string[], c: ICategory) => [...acc, c.name], []);
    },

    // This returns set of selected category
    get selected(): Set<number> {
      const list: Set<number> = new Set();
      self.items.forEach((c: ICategory) => {
        if (c.selected) {
          list.add(c.id);
        }
      });
      return list;
    }
  }))

  .actions((self) => ({
    add(item: ICategory) {
      self.items.put(item);
    },

    setLoading(l: boolean) {
      self.isLoading = l;
    },

    clearSelected() {
      self.items.forEach((c) => {
        c.selected = false;
      });
    },

    toggleByName(name: string) {
      self.items.forEach((c) => {
        if (c.name === name) {
          c.selected = true;
        }
      });
    }
  }))

  .actions((self) => ({
    load: flow(function* () {
      try {
        self.setLoading(true);
        const { api } = self;

        const json = yield api.categories();

        // creating the model only after the store has the tags normalized
        const categories: ICategory[] = json.data.map((c: ICategory) => ({
          id: c.id,
          name: c.name
        }));

        categories.forEach((c: ICategory) => self.add(c));
      } catch (error) {
        const err = error as AxiosError;
        self.err = err.toString();
      }

      self.setLoading(false);
    })
  }))

  .actions((self) => ({
    afterCreate() {
      self.load();
    }
  }));
