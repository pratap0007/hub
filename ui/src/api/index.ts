import axios from 'axios';
import { API_URL } from '../config/constants';
import { ICategory } from '../store/category';
import { IResource, IVersion } from '../store/resource';

interface Token {
  token: string;
  refreshInterval: string;
  expiresAt: number;
}

interface TokenData {
  access: Token;
  refresh: Token;
}

export interface AuthResponse {
  data: TokenData;
}

export interface Api {
  categories(): Promise<ICategory>;
  resources(): Promise<IResource>;
  resourceVersion(resourceId: number): Promise<IVersion>;
  versionUpdate(versionId: number): Promise<IVersion>;
  authentication(authCode: string): Promise<AuthResponse>;
}

export class Hub implements Api {
  async categories() {
    try {
      return axios.get(`${API_URL}/categories`).then((response) => response.data);
    } catch (err) {
      return err.response;
    }
  }

  async resources() {
    try {
      return axios.get(`${API_URL}/resources`).then((response) => response.data);
    } catch (err) {
      return err.response;
    }
  }

  async authentication(authCode: string) {
    try {
      return axios.post(`${API_URL}/auth/login?code=${authCode}`).then((response) => response.data);
    } catch (err) {
      return err.response;
    }
  }

  async resourceVersion(resourceId: number) {
    try {
      return axios
        .get(`${API_URL}/resource/${resourceId}/versions`)
        .then((response) => response.data);
    } catch (err) {
      return err.response;
    }
  }

  async versionUpdate(versionId: number) {
    try {
      return axios
        .get(`${API_URL}/resource/version/${versionId}`)
        .then((response) => response.data);
    } catch (err) {
      return err.response;
    }
  }
}
