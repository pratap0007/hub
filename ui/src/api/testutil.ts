import { Api } from './';
import * as fs from 'fs';
import { ICategory } from '../store/category';

export class FakeHub implements Api {
  dataDir: string;

  constructor(dataDir: string) {
    this.dataDir = dataDir;
  }

  async categories() {
    const data = `${this.dataDir}/categories.json`;

    const ret = () => JSON.parse(fs.readFileSync(data).toString());
    return new Promise<ICategory>((resolve) => {
      setTimeout(() => resolve(ret()), 1000);
    });
  }
}
