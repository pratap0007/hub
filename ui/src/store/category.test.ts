import { when } from 'mobx';
import { getSnapshot } from 'mobx-state-tree';
import { FakeHub } from '../api/testutil';
import { CategoryStore, Category, Tag } from './category';
import { assert } from './utils';

const TESTDATA_DIR = `${__dirname}/testdata`;
const api = new FakeHub(TESTDATA_DIR);

describe('Store Object', () => {
  it('can create a tag object', () => {
    const store = Tag.create({
      id: 1,
      name: 'cli'
    });
    expect(store.name).toBe('cli');
  });
  it('can create a category object', () => {
    const category = Category.create({
      id: 1,
      name: 'test',
      tags: ['1']
    });

    expect(category.name).toBe('test');
    expect(category.tags.length).toBe(1);
  });
});

describe('Store functions', () => {
  it('can create a category store', (done) => {
    const store = CategoryStore.create({}, { api });
    expect(store.count).toBe(0);
    expect(store.isLoading).toBe(true);
    when(
      () => !store.isLoading,
      () => {
        expect(store.count).toBe(5);
        expect(store.isLoading).toBe(false);

        expect(getSnapshot(store)).toMatchSnapshot();

        done();
      }
    );
  });

  it('can toggle the selected category', (done) => {
    const store = CategoryStore.create({}, { api });
    expect(store.count).toBe(0);
    expect(store.isLoading).toBe(true);

    when(
      () => !store.isLoading,
      () => {
        expect(store.count).toBe(5);
        expect(store.isLoading).toBe(false);

        const categories = store.items.get('1');
        assert(categories);
        categories.toggle();

        expect(categories.selected).toBe(true);

        done();
      }
    );
  });

  it('can clear all the categories', (done) => {
    const store = CategoryStore.create({}, { api });
    expect(store.count).toBe(0);

    when(
      () => !store.isLoading,
      () => {
        expect(store.count).toBe(5);
        expect(store.isLoading).toBe(false);

        // Gets the category with id as 1
        const c1 = store.items.get('1');
        assert(c1);
        c1.toggle();

        // Gets the category with id as 2
        const c2 = store.items.get('2');
        assert(c2);
        c2.toggle();

        store.clearSelected();
        expect(c1.selected).toBe(false);

        done();
      }
    );
  });

  it('can return the tags for the categories which are selected', (done) => {
    const store = CategoryStore.create({}, { api });
    expect(store.count).toBe(0);
    expect(store.isLoading).toBe(true);

    when(
      () => !store.isLoading,
      () => {
        expect(store.count).toBe(5);
        expect(store.isLoading).toBe(false);

        // Gets the category with id as 1
        const c1 = store.items.get('1');
        assert(c1);
        c1.toggle();

        // Gets the category with id as 2
        const c2 = store.items.get('2');
        assert(c2);
        c2.toggle();

        expect(store.selectedTags.size).toBe(2);

        done();
      }
    );
  });

  it('can toggle the category by name', (done) => {
    const store = CategoryStore.create({}, { api });
    expect(store.count).toBe(0);
    expect(store.isLoading).toBe(true);

    when(
      () => !store.isLoading,
      () => {
        expect(store.count).toBe(5);
        expect(store.isLoading).toBe(false);

        store.toggleByName('Build Tools');

        const categories = store.items.get('1');
        assert(categories);
        expect(categories.selected).toBe(true);

        done();
      }
    );
  });

  it('can return the all selected catgories in a list', (done) => {
    const store = CategoryStore.create({}, { api });
    expect(store.count).toBe(0);
    expect(store.isLoading).toBe(true);

    when(
      () => !store.isLoading,
      () => {
        expect(store.count).toBe(5);
        expect(store.isLoading).toBe(false);

        // Gets the category with id as 1
        const c1 = store.items.get('1');
        assert(c1);
        c1.toggle();

        // Gets the category with id as 2
        const c2 = store.items.get('2');
        assert(c2);
        c2.toggle();

        expect(store.selectedByName.sort()).toEqual(['Build Tools', 'CLI'].sort());

        done();
      }
    );
  });
});
