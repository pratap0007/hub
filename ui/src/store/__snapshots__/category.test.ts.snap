// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Store functions can create a category store 1`] = `
Object {
  "err": "",
  "isLoading": false,
  "items": Object {
    "1": Object {
      "id": 1,
      "name": "Build Tools",
      "selected": false,
    },
    "2": Object {
      "id": 2,
      "name": "CLI",
      "selected": false,
    },
    "3": Object {
      "id": 3,
      "name": "Cloud",
      "selected": false,
    },
    "4": Object {
      "id": 4,
      "name": "Deploy",
      "selected": false,
    },
    "5": Object {
      "id": 5,
      "name": "Image Build",
      "selected": false,
    },
  },
}
`;
